compiler/luci/pass/src/FoldShapePass.test.cpp

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2024 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/FoldShapePass.h"
+#include "PassTestGraphs.h"
+
+#include <luci/IR/CircleNodes.h>
+
+#include <gtest/gtest.h>
+
+namespace
+{
+
+template <loco::DataType OutType> class FoldShapeGraph : public luci::ConstantFoldingAddTestGraph
+{
+public:
+  FoldShapeGraph(std::vector<uint32_t> input_shape)
+    : luci::ConstantFoldingAddTestGraph(input_shape, OutType)
+  {
+    _x = _g.nodes()->template create<luci::CircleConst>();
+    _x->name("x");
+    _x->dtype(loco::DataType::S32);
+    _x->rank(input_shape.size());
+    for (uint32_t i = 0; i < input_shape.size(); i++)
+      _x->dim(i).set(input_shape.at(i));
+    _x->shape_status(luci::ShapeStatus::VALID);
+
+    _shape = _g.nodes()->template create<luci::CircleShape>();
+    _shape->name("shape");
+    _shape->out_type(OutType);
+    _shape->input(_x);
+    _shape->shape({4});
+    _shape->rank(1);
+    _shape->dim(0).set(4);
+  }
+
+  loco::Node *createFoldedPattern() override { return _shape; }
+
+protected:
+  luci::CircleShape *_shape = nullptr;
+  luci::CircleConst *_x = nullptr;
+};
+
+/**
+ *  Graph that has a Shape Op
+ *
+ *    BEFORE
+ *
+ *    [CircleInput]   [CircleShape]
```

Comment:
`[CircleShape]` has a `CircleConst` as input

Reply:
I think that you want to say same thing from https://github.com/Samsung/ONE/pull/12478#discussion_r1462870644

I think what you said means the tensor whose shape is `[1, 8, 8, 64]` is input of `[CircleShape]` (just in this test code..!)

This is true, but I'm bit confused because it is hard to decide that which diagram is more proper to explain the graph

I think there are two possible options. Please let me know how do you think about it 😄

---
### option 1
```cpp
/**
 *  Graph that has a Shape Op
 *
 *    BEFORE
 *
 *                    [CircleConst]
 *                          |
 *    [CircleInput]   [CircleShape]
 *              \      /
 *             [CircleAdd]
 *                 |
 *            [CircleOutput]
 *
 *    AFTER
 *
 *    [CircleInput]   [CircleConst]
 *              \      /
 *             [CircleAdd]
 *                 |
 *            [CircleOutput]
 *
 */
```

---
### option 2
```cpp
/**
 *  Graph that has a Shape Op
 *
 *    BEFORE
 *
 *             [CircleInput]
 *                 |      \
 *                 |    [CircleShape]
 *                 |      /
 *             [CircleAdd]
 *                 |
 *            [CircleOutput]
 *
 *    AFTER
 *
 *    [CircleInput]   [CircleConst]
 *              \      /
 *             [CircleAdd]
 *                 |
 *            [CircleOutput]
 *
 */
```

---
IMHO, second options looks more reasonable due to below reaseons

- The input of `Shape` and `Add` is identical (at here, `[1, 8, 8, 64]`)
- According to https://github.com/Samsung/ONE/pull/12478#discussion_r1465910319, pass itself does not remove any node
  > To be exact, all the pass don't remove the node, just changes the connection.

  But, if we choose first option, the `[CircleShape]` node disappear
  Or, to avoid this, the `[CircleShape]` node should be marked as dangling node, but it looks bit strange (because `[CircleShape]` was folded and was replaced as `[CircleConst]`...) and I think that it can cause confusion


Consequently, I updated the comment as following the second option now
But if there is any error or something I misunderstood, please let me know :)

Reply:
Q) how can the graph of test, `FoldShapeGraph`, you wrote can have two options?

Reply:
Actually, that is the confusing point to me... 😥

In this test, the input of the graph is constant, and it works as a input of `CircleShape`

So I thought like that the input of the `[CircleShape]` can be marked as `[CircleInput]` and `[CircleConst]` both, thus I thought that there are two options

To **option 1**, the input of `[CircleShape]` was marked as `[CircleConst]`
To **option 2**, the input of `[CircleShape]` was marked as `[CircleInput]`

but I'm not sure which one is right...

Reply:
> To option 1, the input of [CircleShape] was marked as [CircleConst]

What do you mean by `marked` ?

> To option 2, the input of [CircleShape] was marked as [CircleInput]

What do you mean by `[CircleInput]` ?

Reply:
> Consequently, I updated the comment as following the second option now

And I've pushed the code without the comment change by mistake 😅
It is updated now!

Reply:
I think you don't understand the diagram in the comment.

Reply:
I've totally misunderstood about it..! 😢

(For note)
After offline talk with @seanshpark, I understood what the comments mean

> `[CircleSomething]` in the diagram is intended to express the circle node IR.

Like already said in https://github.com/Samsung/ONE/pull/12478?notification_referrer_id=NT_kwDOAq4Po7M5MDkzNDY2MDgwOjQ0OTYxNjk5#discussion_r1465910319, `[CircleSomething]` is a Circle IR node, and it is implemented as a class at the code

It means that there are a classes named `CircleInput`/`CircleOutput`/`CircleShape`/`CircleConst` and so on

So the nodes which were commented here represent each correspond node for the test graph!